### DISCUSSION TITLE
handling hidden imports which are only available on some targets

### DISCUSSION THREAD

**Discussion by maxbachmann** [pyinstaller/pyinstaller#8641] [Help] [answered]
In one of my libraries I do provide a pyinstaller hook to list hidden imports. As far as I understood it's preferable to list only hidden imports which actually exist so there are no warnings about hidden imports.

1) how can I test there are no such warnings? Even if I list a hidden import which doesn't exist this isn't reported when running the hook unit tests using  `python -m PyInstaller.utils.run_tests --include_only myLibrary.`

2) I have some hidden imports which only exist for specific hardware + os combinations like windows x64. How should those be listed in the hiddenimports? I assume I can't just ask Python for it since the packaging tool wouldn't necessarily run on the target os.

**Comment by bwoodsend** [comment, marked as answer]
Are these hidden imports your own submodules or external modules that they depend on?

> I have some hidden imports which only exist for specific hardware + os combinations like windows x64. How should those be listed in the hiddenimports? I assume I can't just ask Python for it since the packaging tool wouldn't necessarily run on the target os.

PyInstaller is not a cross compiler so conditionals around `platform.system()` and `platform.machine()` are fine.

> As far as I understood it's preferable to list only hidden imports which actually exist so there are no warnings about hidden imports.

It is in the backs of our minds that we need to change the warnings for missing hidden inputs. I think we'll eventually make it that any missing hidden import from a hook is a silent no-op

Assuming that by hidden imports not existing you really do mean they're not there and not that they merely raise exceptions, [collect_submodules()](https://pyinstaller.org/en/stable/hooks.html#PyInstaller.utils.hooks.collect_submodules) will give you the list of only submodules that exist.

**Reply by maxbachmann** [reply]
> Are these hidden imports your own submodules or external modules that they depend on?

The hidden imports are all submodules.

> Assuming that by hidden imports not existing you really do mean they're not there and not that they merely raise exceptions, [collect_submodules()](https://pyinstaller.org/en/stable/hooks.html#PyInstaller.utils.hooks.collect_submodules) will give you the list of only submodules that exist.

Yes they really don't exist. I wasn't aware of this utility function and this certainly simplifies the maintenance of these hidden imports :+1:
